srcs/HttpRequest/HttpRequest/HttpRequest.cpp

```diff
@@ -0,0 +1,1305 @@
+		if (is_requestformat(line) == false)
+		{
+			this->_status_code = 400;
+			return;
+		}
+		if (HandlingString::is_printablecontent(remove_sohword_line) == false)
+		{
+			this->_status_code = 400;
+			return;
+		}
+		key = this->obtain_request_key(line);
+		value = this->obtain_request_value(line);
+		value = value.substr(0, value.length() - 1);
+		if (this->is_keyword_exist(key) == true)
+			(this->*_inputvalue_functionmap[key])(key, value);
+	}
+}
+
+HttpRequest::~HttpRequest()
+{
+	std::map<std::string, BaseKeyValueMap*>::iterator inputed_class_itr = this->_request_keyvalue_map.begin();
+	while (inputed_class_itr != this->_request_keyvalue_map.end())
+	{
+		delete (inputed_class_itr->second);
+		inputed_class_itr++;
+	}
+}
+
+LinkClass* HttpRequest::ready_LinkClass(std::map<std::string, std::map<std::string, std::string> > link_valuemap)
+{
+	return (new LinkClass(link_valuemap));
+}
+
+SecurityPolicy* HttpRequest::ready_SecurityPolicy(const std::string &report_url, std::map<std::string, std::vector<std::string> >	_policy_directive)
+{
+	return (new SecurityPolicy(report_url, _policy_directive));
+}
+
+SecurityPolicy* HttpRequest::ready_SecurityPolicy(std::map<std::string, std::vector<std::string> >	_policy_directive)
+{
+	return (new SecurityPolicy(_policy_directive));
+}
+
+TwoValueSet* HttpRequest::ready_TwoValueSet(const std::string &all_value)
+{
+	std::stringstream	ss(HandlingString::obtain_value(all_value));
+	std::string			first_value;
+	std::string			second_value;
+
+	std::getline(ss, first_value, '/');
+	std::getline(ss, second_value, '/');
+
+	return (new TwoValueSet(first_value, second_value));
+}
+
+TwoValueSet* HttpRequest::ready_TwoValueSet(const std::string &value, char delimiter)
+{
+	std::stringstream	ss(HandlingString::obtain_value(value));
+	std::string			first_value;
+	std::string			second_value;
+
+	std::getline(ss, first_value, delimiter);
+	std::getline(ss, second_value, delimiter);
+	return (new TwoValueSet(HandlingString::obtain_value(first_value), HandlingString::obtain_value(second_value)));
+}
+
+ValueArraySet* HttpRequest::ready_ValueArraySet(const std::string &all_value)
+{
+	std::vector<std::string>	value_array;
+
+	std::stringstream	ss(all_value);
+	std::string			line;
+	while(std::getline(ss, line, ','))
+		value_array.push_back(HandlingString::obtain_value(line));
+	return (new ValueArraySet(value_array));
+}
+
+ValueDateSet* HttpRequest::ready_ValueDateSet(const std::string &value)
+{
+	return (new ValueDateSet(HandlingString::obtain_value(value)));
+}
+
+ValueMap* HttpRequest::ready_ValueMap(const std::string &value, char delimiter)
+{
+	std::map<std::string, std::string> value_map;
+	std::stringstream	ss(value);
+	std::string			line;
+
+	while(std::getline(ss, line, delimiter))
+		value_map[HandlingString::obtain_word_beforedelimiter(HandlingString::skip_emptyword(line), '=')] \
+		= HandlingString::obtain_afterword(HandlingString::obtain_value(line), '=');
```

Comment:
```suggestion
	while(std::getline(ss, line, delimiter)) {
		std::string space_removed_line = HandlingString::skip_emptyword(line);
		std::string field_name = HandlingString::obtain_word_beforedelimiter(space_removed_line, '=');
		std::string field_value_with_ows = HandlingString::obtain_value(line);
		std::string field_value = HandlingString::obtain_afterword(field_value_with_ows, '=');

		value_map[field_name] = field_value;
	}
```

MUST:
- [ ] これくらいに分解して欲しい🥺
関数のネストは読んでいて辛いです
特にC++のように名前空間があると情報量が多いので把握しにくすぎる

Reply:
ここまで分解するのかを迷います
関数名をわかりやすくしているのはこういった
コードを横に一列に記載できるようにする、という印象がありました

Reply:
https://www.google.co.jp/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwj5pIf2sMGBAxWWslYBHfojDcwQFnoECA8QAQ&url=https%3A%2F%2Fzenn.dev%2Fn_harada%2Farticles%2Freadability-guideline&usg=AOvVaw0apfEvIYbq-GgEtehsATRT&opi=89978449

Reply:
一回しか使わないものを変数に置き換えるというところがあまり納得できていません。。。

Reply:
関数名をわかりやすくしているのは、主に可読性、保守性などの観点です。
実態に合わせた関数名にすることで、処理を把握しやすいので、不具合などに気づきやすくなります。

コードを横に一列に記載できるようにする狙いはなかったです。
（個人的には長い1行よりも、複数行に分割された方が見やすいと思っている）

<br>

> 一回しか使わないものを変数に置き換えるというところがあまり納得できていません。。。

上と下、意図を正確に把握し、挙動が追いやすいのはどちらだと思いますか？
また、例えば、誤って`obtain_word_beforedelimiter()`が`obtain_afterword()`になっていたとして、気づきやすいのはどちらだと思いますか？

```c++
value_map[HandlingString::obtain_word_beforedelimiter(HandlingString::skip_emptyword(line), '=')] \
		= HandlingString::obtain_afterword(HandlingString::obtain_value(line), '=');

--------

std::string space_removed_line = HandlingString::skip_emptyword(line);
std::string field_name = HandlingString::obtain_word_beforedelimiter(space_removed_line, '=');
std::string field_value_with_ows = HandlingString::obtain_value(line);
std::string field_value = HandlingString::obtain_afterword(field_value_with_ows, '=');

value_map[field_name] = field_value;
```

Reply:
やっぱりこの話ってローコンテキストパターンの話でよかったのですね
ふんふん。。。

Reply:
これって関数名が長くなってしまってるからという観点も入っていますか？（もしくは関数の引数で別の関数を呼び出している

Reply:
めっちゃ簡単な例ですが
compare(sum(a,b), sum(c,d))
みたいな関数があれば
aとbの合計値とcとdの合計値を比較するんや
みたいな判定になるかと思います

Reply:
> これって関数名が長くなってしまってるからという観点も入っていますか？（もしくは関数の引数で別の関数を呼び出している

それもあります。
```
func_a(func_b(func_c, A), func_d));
```
みたいにネストすると、何やってるか分かりにくいと思います。

HandlingStringの名前空間が入ってきているので、より冗長に感じるのかも



> めっちゃ簡単な例ですが compare(sum(a,b), sum(c,d)) みたいな関数があれば aとbの合計値とcとdの合計値を比較するんや みたいな判定になるかと思います

sum()みたいなシンプルでreturn valueが明確なら、これでも良いと思います。

Reply:
この場合って
a_b_sum = sum(a, b)
c_d_sum = sum(cud)
compare(a_b_sum, c_d_sum)
みたいな感じで描くのはちょっと冗長になってそうという印象を受けました
（冗長じゃなかったら申し訳ないです

Reply:
みたいに考えるとある一定の関数名の長さを超えた場合は別途わかりやすい単語に置き換える等の処理をする必要がある、みたいな判断なのかなと思っています

Reply:
> これって関数名が長くなってしまってるからという観点も入っていますか？（もしくは関数の引数で別の関数を呼び出している

それもあります。
```
func_a(func_b(func_c, A), func_d));
```
みたいにネストすると、何やってるか分かりにくいと思います。

HandlingStringの名前空間が入ってきているので、より冗長に感じるのかも

<br>

> めっちゃ簡単な例ですが compare(sum(a,b), sum(c,d)) みたいな関数があれば aとbの合計値とcとdの合計値を比較するんや みたいな判定になるかと思います

sum()みたいなシンプルでreturn valueが明確なら、これでも良いと思います。

<br>

> この場合って a_b_sum = sum(a, b) c_d_sum = sum(cud) compare(a_b_sum, c_d_sum) みたいな感じで描くのはちょっと冗長になってそうという印象を受けました （冗長じゃなかったら申し訳ないです

おっしゃるように、これは冗長かもしれ右脳ませんね。
sum()をの返り値を変数にか冗長に感じるのは、
* 返り値が明確で、sum(a,b)の記述で十分だから
* この例ではa, bなどの変数名なので、実運用と異なる例だから
な気がします。

今、↑のHandlingStringの関数では、
もちゃさんが`sum(a,b)`と思っているのかもしれませんが

私には`sum_and_nanka(a, nanka_suru_kansu(b, c))`みたいに見えています。
HandlingStringの名前空間が入ってきているので、より冗長に感じている感はありますが...

<br>

> みたいに考えるとある一定の関数名の長さを超えた場合は別途わかりやすい単語に置き換える等の処理をする必要がある、みたいな判断なのかなと思っています

関数の設計次第、な気もします。

Reply:
> これって関数名が長くなってしまってるからという観点も入っていますか？

関数名を省略しすぎる・抽象的すぎるよりも、多少長くても具体的な方がいいと思います（リーダブルコード p.16）。

<br>

> もしくは関数の引数で別の関数を呼び出している

こっちかもです。
```
func_a(func_b(func_c, A), func_d));
```
みたいにネストすると、何をやってるか分かりにくいと思います。
（これはテキトーに書いた例ですが...）

<br>

> めっちゃ簡単な例ですが compare(sum(a,b), sum(c,d)) みたいな関数があれば aとbの合計値とcとdの合計値を比較するんや みたいな判定になるかと思います
> この場合って a_b_sum = sum(a, b) c_d_sum = sum(cud) compare(a_b_sum, c_d_sum) みたいな感じで描くのはちょっと冗長になってそうという印象を受けました （冗長じゃなかったら申し訳ないです

`sum()`みたいなシンプルでreturn valueが明確なら、これでも良いと思います。

おっしゃるように、`sum()`の返り値を変数に格納するのは冗長かもしれませんね。
これが冗長だと感じるのは、
* `sum(a,b)`が十分実態を表している
* この例では`a`, `b`などの変数名で意味を持っておらず、実運用と異なる例だから

な気がします。
`a`, `b`がそれぞれなんらかの意味をもっており、`sum(a, b)`が特別な意味を持つような演算なら、
`sum(a,b)`の結果を適当な変数名に格納した後に`compare()`した方が良いかもしれませんね。
これは、`sum(a,b)`の結果が持つ重要な意味を握りつぶして、`compare()`に渡しているためで、
`sum(a,b)`の意味を知らない人には`a, bの合計値`以上の意味はなく、その意図を理解する手段を絶った実装になってしまうためです。

<br>

> みたいに考えるとある一定の関数名の長さを超えた場合は別途わかりやすい単語に置き換える等の処理をする必要がある、みたいな判断なのかなと思っています

関数の設計次第、な気もします。
関数が実態を十分表していて、可読性を損なわなければ変数に格納するなどの必要はないと思います。

<br>

↓の 右辺は、`HandlingString::obtain_value(line), '=');`の結果が`field_value_with_ows`になると思いますが、
それが暗黙的になっており、実装者以外が読み取ることは難しい気がします。

`obtain_afterword(obtain_value(line), '=');`でobtainできるものが、
`=`の後の文字列であることはなんとなく分かりますが、
その文字列が持つ役割を読み解くには、コードをひたすら読んでいくしかありません。

ちなみに、関数ジャンプなどでいろいろ飛びましたが、
`value_map`が`map<string, string>`であること以外はわからず、
`key = field-name`, `value = field-value`であることを読み解くのは難しいかもしれません。
```c++
value_map[HandlingString::obtain_word_beforedelimiter(HandlingString::skip_emptyword(line), '=')] \
		= HandlingString::obtain_afterword(HandlingString::obtain_value(line), '=');

value_map[obtain_word_beforedelimiter(skip_emptyword(line), '=')] \
		= obtain_afterword(obtain_value(line), '=');
```

<br>

関数名が`HandlingString::obtain_value_woth_owd(line), '=');`ならまだ分かります。
この辺りの関数はライブラリとして作成されていると思うので、ドンピシャな命名が難しいと思いますが、その際に一時的に変数に格納する（説明変数のような役割）があると、関数を呼び出している意図や、その結果受け取りたいものが何であるかが明確になると思います。
```c++
std::string field_value_with_ows = HandlingString::obtain_value(line);
std::string field_value = HandlingString::obtain_afterword(field_value_with_ows, '=');
```

<br>

また、関数名が`sum()`のように実態を直接表していれば、直接記述しても違和感は覚えません。

```
value_map[HandlingString::obtain_word_beforedelimiter(HandlingString::skip_emptyword(line), '=')] \
		= HandlingString::obtain_afterword(HandlingString::obtain_value(line), '=');

value_map[HandlingString:get_field_name(line)] = HandlingString::get_field_value(line));
```

Reply:
なるほど